tests/timeseries/_formula_engine/test_formula_composition.py

```diff
@@ -41,33 +42,56 @@ async def test_formula_composition(  # pylint: disable=too-many-locals
         engine = (logical_meter.pv_power + battery_pool.power).build("inv_power")
         inv_calc_recv = engine.new_receiver()
 
-        count = 0
-        for _ in range(10):
-            grid_pow = await grid_power_recv.receive()
-            pv_pow = await pv_power_recv.receive()
-            bat_pow = await battery_power_recv.receive()
-            main_pow = await main_meter_recv.receive()
-            inv_calc_pow = await inv_calc_recv.receive()
-
-            assert grid_pow is not None and grid_pow.value is not None
-            assert inv_calc_pow is not None and inv_calc_pow.value is not None
-            assert bat_pow is not None and bat_pow.value is not None
-            assert pv_pow is not None and pv_pow.value is not None
-            assert main_pow is not None and main_pow.value is not None
-            assert isclose(inv_calc_pow.value, pv_pow.value + bat_pow.value)
-            assert isclose(grid_pow.value, inv_calc_pow.value + main_pow.value)
-            count += 1
+        await mockgrid.mock_data.send_bat_inverter_power([10.0, 12.0, 14.0])
+        await mockgrid.mock_data.send_meter_power(
+            [100.0, 10.0, 12.0, 14.0, -100.0, -200.0]
+        )
+
+        grid_pow = await grid_power_recv.receive()
+        pv_pow = await pv_power_recv.receive()
+        bat_pow = await battery_power_recv.receive()
+        main_pow = await main_meter_recv.receive()
+        inv_calc_pow = await inv_calc_recv.receive()
+
+        assert (
+            grid_pow is not None
+            and grid_pow.value is not None
+            and math.isclose(grid_pow.value, -164.0)
+        )  # 100 + 10 + 12 + 14 + -100 + -200
+        assert (
+            bat_pow is not None
+            and bat_pow.value is not None
+            and math.isclose(bat_pow.value, 36.0)
+        )  # 10 + 12 + 14
+        assert (
+            pv_pow is not None
+            and pv_pow.value is not None
+            and math.isclose(pv_pow.value, -300.0)
+        )  # -100 + -200
+        assert (
+            inv_calc_pow is not None
+            and inv_calc_pow.value is not None
+            and math.isclose(inv_calc_pow.value, -264.0)  # -300 + 36
+        )
+        assert (
+            main_pow is not None
+            and main_pow.value is not None
+            and math.isclose(main_pow.value, 100.0)
```

Comment:
Alternatively we can use [pytest.approx](https://docs.pytest.org/en/7.1.x/reference/reference.html#id8) in tests. This is just an optional comment...